ast/DuplicateHashKeyCheck.h

```diff
@@ -0,0 +1,54 @@
+#ifndef SORBET_AST_DUPLICATE_HASH_KEY_CHECK_H
+#define SORBET_AST_DUPLICATE_HASH_KEY_CHECK_H
+
+#include "ast/Trees.h"
+#include "common/common.h"
+#include "core/errors/desugar.h"
+
+namespace sorbet::ast {
+
+class DuplicateHashKeyCheck {
+    const core::MutableContext ctx;
+    UnorderedMap<core::NameRef, core::LocOffsets> hashKeySymbols;
+    UnorderedMap<core::NameRef, core::LocOffsets> hashKeyStrings;
+
+public:
+    DuplicateHashKeyCheck(const core::MutableContext &ctx) : ctx{ctx}, hashKeySymbols(), hashKeyStrings() {}
+
+    void check(const ExpressionPtr &key) {
+        auto lit = ast::cast_tree<ast::Literal>(key);
+        if (lit == nullptr || !lit->isName()) {
+            return;
+        }
+        auto nameRef = lit->asName();
+
+        auto isSymbol = lit->isSymbol();
+        if (isSymbol && !hashKeySymbols.contains(nameRef)) {
+            hashKeySymbols[nameRef] = key.loc();
+        } else if (!isSymbol && !hashKeyStrings.contains(nameRef)) {
+            hashKeyStrings[nameRef] = key.loc();
+        } else {
+            if (auto e = ctx.beginIndexerError(key.loc(), core::errors::Desugar::DuplicatedHashKeys)) {
+                core::LocOffsets originalLoc = isSymbol ? hashKeySymbols[nameRef] : hashKeyStrings[nameRef];
+
+                e.setHeader("Hash key `{}` is duplicated", nameRef.toString(ctx.state));
+                e.addErrorLine(ctx.locAt(originalLoc), "First occurrence of `{}` hash key",
+                               nameRef.toString(ctx.state));
+            }
+        }
+    }
+
+    // This is only used with Send::ARGS_store and Array::ELEMS_store
+    template <typename T> static void checkSendArgs(const core::MutableContext ctx, int numPosArgs, const T &args) {
```

Comment:
This should not be generic. This should just take an `absl::Span<const ExpressionPtr>`